### PR TITLE
[Cocoa] Register Alternate WebM Player skeleton

### PR DESCRIPTION
**Pull request by youssefsoli**
#### bfe3f272a5b8cc261b68249303d357fd424ec572
<pre>
[Cocoa] Register Alternate WebM Player skeleton
<a href="https://bugs.webkit.org/show_bug.cgi?id=241769">https://bugs.webkit.org/show_bug.cgi?id=241769</a>
&lt;rdar://95542551&gt;

Patch by Youssef Soliman &lt;youssefdevelops@gmail.com&gt; on 2022-06-22
Reviewed by Eric Carlson.

Added the alternate WebM player skeleton and registered it behind
the internal feature flag.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/RuntimeEnabledFeatures.cpp:
(WebCore::RuntimeEnabledFeatures::setAlternateWebMPlayerEnabled):
* Source/WebCore/page/RuntimeEnabledFeatures.h:
(WebCore::RuntimeEnabledFeatures::alternateWebMPlayerEnabled const):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::setAlternateWebMPlayerEnabled):
(WebCore::PlatformMediaSessionManager::alternateWebMPlayerEnabled):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::WTF_REQUIRES_LOCK):
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h: Added.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm: Added.
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::~MediaPlayerPrivateWebM):
(WebCore::mimeTypeCache):
(WebCore::MediaPlayerPrivateWebM::getSupportedTypes):
(WebCore::MediaPlayerPrivateWebM::supportsType):
(WebCore::MediaPlayerPrivateWebM::load):
(WebCore::MediaPlayerPrivateWebM::cancelLoad):
(WebCore::MediaPlayerPrivateWebM::platformLayer const):
(WebCore::MediaPlayerPrivateWebM::play):
(WebCore::MediaPlayerPrivateWebM::pause):
(WebCore::MediaPlayerPrivateWebM::setPageIsVisible):
(WebCore::MediaPlayerPrivateWebM::currentMediaTime const):
(WebCore::MediaPlayerPrivateWebM::buffered const):
(WebCore::MediaPlayerPrivateWebM::didLoadingProgress const):
(WebCore::MediaPlayerPrivateWebM::logChannel const):
(WebCore::MediaPlayerPrivateWebM::registerMediaEngine):

Changes were done to the GPUProcessPreferences object to consolidate
all changes when adding a new feature to just one file.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::updateGPUProcessPreferences):
* Source/WebKit/GPUProcess/GPUProcessPreferences.cpp:
(WebKit::GPUProcessPreferences::GPUProcessPreferences):
(WebKit::GPUProcessPreferences::copyEnabledWebPreferences):
(WebKit::GPUProcessPreferences::encode const):
(WebKit::GPUProcessPreferences::decode):
* Source/WebKit/GPUProcess/GPUProcessPreferences.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::updatePreferences):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: <a href="https://commits.webkit.org/251767@main">https://commits.webkit.org/251767@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295762">https://svn.webkit.org/repository/webkit/trunk@295762</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Source/WebCore/WebCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -24765,15 +24767,15 @@
...
 				6FB17475283A8FF40067D8CA /* FlexFormattingConstraints.h */,
 				6FB7D2D7250FD7E5000207AA /* FlexFormattingContext.cpp */,
 				6FB7D2D9250FD7FC000207AA /* FlexFormattingContext.h */,
 				6F70DEDD251126F300F0FC78 /* FlexFormattingGeometry.cpp */,
 				6FF911F626487FC8002021DF /* FlexFormattingGeometry.h */,
 				6FB7D2D8250FD7EF000207AA /* FlexFormattingState.cpp */,
 				6FB7D2DA250FD7FC000207AA /* FlexFormattingState.h */,
+				6F047A9028453ED200C25EE7 /* FlexLayout.cpp */,
+				6F047A9128453EDB00C25EE7 /* FlexLayout.h */,
 				6FB9105B2830BC8A00004929 /* FlexRect.h */,
```
Comment (youssefsoli) [inline Source/WebCore/WebCore.xcodeproj/project.pbxproj]: These unrelated changes are due to the automatic sorting of modified `.pbxproj` files.

File: Source/WebKit/WebProcess/WebPage/WebPage.cpp [hunk 2]
```diff
@@ -940,6 +940,9 @@ WebPage::WebPage(PageIdentifier pageID, WebPageCreationParameters&& parameters)
 #if ENABLE(GPU_PROCESS)
 void WebPage::gpuProcessConnectionDidBecomeAvailable(GPUProcessConnection& gpuProcessConnection)
 {
+#if ENABLE(ALTERNATE_WEBM_PLAYER)
+    gpuProcessConnection.setAlternateWebMPlayerEnabled(DeprecatedGlobalSettings::isAlternateWebMPlayerEnabled());
```
Comment (youssefsoli) [inline Source/WebKit/WebProcess/WebPage/WebPage.cpp]: There might be a better place to put this update rather than in the `gpuProcessConnectionDidBecomeAvailable` event?

File: Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml [hunk 3]
```diff
@@ -58,14 +58,14 @@ AlternateWebMPlayerEnabled:
   type: bool
   humanReadableName: "Alternate WebM Player"
   humanReadableDescription: "Enable Alternate WebM Player"
+  getter: isAlternateWebMPlayerEnabled
+  webcoreBinding: DeprecatedGlobalSettings
```
Comment (weinig) [inline Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml]: Please don't add more to 'DeprecatedGlobalSettings'. The name is meant to make it clear this is something that should not be used.

Instead, all new settings should live on WebCore::Settings, and if lower levels of the stack need them, delegation or other means can be used to make them available.

**Review by weinig — CHANGES_REQUESTED** [review]

File: Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml (hunk 3 above)
Comment (youssefsoli) [reply Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml]: How do you recommend exposing these settings to a static context? (in the static buildMediaEnginesVector function)

File: Source/WebCore/WebCore.xcodeproj/project.pbxproj (hunk 1 above)
Comment (youssefsoli) [reply Source/WebCore/WebCore.xcodeproj/project.pbxproj]: Resolved through https://commits.webkit.org/251668@main

File: Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml (hunk 3 above)
Comment (youssefsoli) [reply Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml]: The alternatives I have thought of have been to use `RuntimeEnabledFeatures` which I understand is also deprecated, or to pass a `WebCore::Settings` object into the static context and into the GPU Process every time the page updates (or I can just pass the individual boolean around instead of the whole object).

The first alternative is very similar to using `DeprecatedGlobalSettings` and the second alternative has issues since the GPU Process is shared amongst different pages (which one gets priority?).

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    MediaPlayer::NetworkState m_networkState { MediaPlayer::NetworkState::Empty };
+    MediaPlayer::ReadyState m_readyState { MediaPlayer::ReadyState::HaveNothing };
+    
+    Ref<const Logger> m_logger;
+    const void* m_logIdentifier;
+    
+    FloatSize m_naturalSize;
+    bool m_hasAudio { false };
+    bool m_hasVideo { false };
+    MediaTime m_currentTime { MediaTime::zeroTime() };
```
Comment (jyavenard) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h]: could use the default constructor. MediaTime() initialise by default to zeroTime()

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+#if ENABLE(ALTERNATE_WEBM_PLAYER)
+
+#include "MediaPlayerPrivate.h"
+#include "PlatformLayer.h"
+#include <wtf/LoggerHelper.h>
+#include <wtf/WeakPtr.h>
+
+namespace WebCore {
+
+class MediaPlayerPrivateWebM
```
Comment (jyavenard) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:37]: MediaPlayerPrivateWebM should be using abstracted method from either SourceBufferPrivate or MediaPlayerPrivate such as it's platform independent and not cocoa only.

Isn't this how your actual implementation works?

File: Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml (hunk 3 above)
Comment (weinig) [reply Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml]: In general, static interfaces should not exist in WebCore. Static interfaces don't work in cases that two pages in the same web process (or gpu process) have different WebPreferences. Beyond that, global state makes it harder to reason about things locally so are good thing to avoid anyway. 

The usual model for something like this in WebCore is to have it be bound to the lifetime of the Page, and use the Settings of the Page object to drive it.

In addition to passing down state, you can also use delegation to have the platform code "ask" the higher level what to do.

Comment (weinig) [reply Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml]: This may require refactoring how the MediaPlayer builds up the engines to not be static itself.

Comment (youssefsoli) [reply Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml]: Good points, I'll discuss this further with the rest of the media team.

**Comment by youssefsoli** [comment]
I filed this bug track the switch to page aware settings for media-related flags https://bugs.webkit.org/show_bug.cgi?id=241821.

For now, I put the flag for this feature beside the rest of the media features so it can get picked up in the future when the bug is resolved.

File: Source/WebCore/page/RuntimeEnabledFeatures.cpp [hunk 6]
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2009 Google Inc. All rights reserved.
- * Copyright (C) 2013-2017 Apple Inc. All rights reserved.
+ * Copyright (C) 2013-2012 Apple Inc. All rights reserved.
```
Comment (eric-carlson) [inline Source/WebCore/page/RuntimeEnabledFeatures.cpp]: Time travel!

File: Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp [hunk 7]
```diff
@@ -28,6 +28,7 @@
 
 #include "AudioSession.h"
 #include "Document.h"
+#include "HTMLMediaElement.h"
```
Comment (eric-carlson) [inline Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp]: This is a layering violation, we can't reach up out of /platform/

File: Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml [hunk 8]
```diff
@@ -58,14 +58,13 @@ AlternateWebMPlayerEnabled:
   type: bool
   humanReadableName: "Alternate WebM Player"
   humanReadableDescription: "Enable Alternate WebM Player"
+  webcoreBinding: RuntimeEnabledFeatures
```
Comment (eric-carlson) [inline Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml]: As we discussed, this should be a Setting.

File: Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp [hunk 9]
```diff
@@ -695,6 +700,27 @@ void PlatformMediaSessionManager::setOpusDecoderEnabled(bool enabled)
 #endif
 }
 
+void PlatformMediaSessionManager::setAlternateWebMPlayerEnabled(bool enabled)
+{
+#if ENABLE(ALTERNATE_WEBM_PLAYER)
+    if (m_alternateWebMPlayerEnabled != enabled) {
+        m_alternateWebMPlayerEnabled = enabled;
+        HTMLMediaElement::resetMediaEngines();
```
Comment (eric-carlson) [inline Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp]: So this will have to be done at the call site, or via a lambda set on `PlatformMediaSessionManager` at some point.

**Review by eric-carlson — APPROVED** [review]
